src/mlib/int128.test.cpp

```diff
@@ -0,0 +1,422 @@
+   CHECK (in_binary == 70917759074396274376598533126648930184_i128);
+   CHECK (
+      in_binary ==
+      "0b110101010110100100001101111001111010100010111100100101101011010110101001010110110011000100000100011110010101101001111110001000"_i128);
+
+   // Or hexadecimal
+   auto in_hex = 0x355a4379ea2f25ad6a56cc411e569f88_i128;
+   CHECK (in_hex == 70917759074396274376598533126648930184_i128);
+
+   int8_t n = -12;
+   CHECK (mlib_int128_scmp (zero, MLIB_INT128_CAST (n)) > 0);
+   CHECK (mlib_int128_ucmp (zero, MLIB_INT128_CAST (n)) < 0);
+
+   auto _2pow127 = mlib_int128_pow2 (127);
+   CHECK (std::string (mlib_int128_format (_2pow127).str) ==
+          "170141183460469231731687303715884105728");
+
+   auto r = div_check (27828649044156246570177174673037165454_i128,
+                       499242349997913298655486252455941907_i128);
+
+   CHECK (r.quotient == 55_i128);
+   CHECK (r.remainder == 370319794271015144125430787960360569_i128);
+
+   r = div_check (64208687961221311123721027584_i128, 3322092839076102144_i128);
+   CHECK (r.remainder == 3155565729965670400_i128);
+
+   // This division will trigger the rare Knuth 4.3.1D/D6 condition:
+   r = div_check (31322872034807296605612234499929458960_i128,
+                  34573864092216774938021667884_i128);
+   CHECK (r.quotient == 905969663_i128);
+   CHECK (r.remainder == 34573864092065898160364055868_i128);
+
+   // Self-divide:
+   r = div_check (628698094597401606590302208_i128,
+                  628698094597401606590302208_i128);
+   CHECK (r.quotient == 1_i128);
+   CHECK (r.remainder == 0_i128);
+
+   // With no high-32 bits in the denominator
+   r = div_check (316356263640858117670580590964547584140_i128,
+                  13463362962560749016052695684_i128);
+   CHECK (r.quotient == 23497566285_i128);
+
+   // Remainder correctness with high bit set:
+   r = div_check (292590981272581782572061492191999425232_i128,
+                  221673222198185508195462959065350495048_i128);
+   CHECK (r.remainder == 70917759074396274376598533126648930184_i128);
+
+   // Remainder with 64bit denom:
+   r = div_check (2795722437127403543495742528_i128, 708945413_i128);
+   CHECK (r.remainder == 619266642_i128);
+
+   // 10-div:
+   CHECK (mlib_int128_div (MLIB_INT128_SMAX, 10_i128) ==
+          17014118346046923173168730371588410572_i128);
+
+   std::random_device rd;
+   std::seed_seq seed ({rd (), rd (), rd (), rd ()});
+   // Pick every numerator bit pattern from 0b00'00 to 0b11'11
+   for (auto nbits = 0u; nbits < 16u; ++nbits) {
+      // This is an extremely rudimentary thread pool to parallelize the
+      // division checks. It doesn't need to be rigorous or optimal, it only
+      // needs to "just work."
+      std::vector<std::thread> threads;
+      threads.resize (15);
+      // Pick every denominator bit pattern from 0b00'01 to 0b11'11:
+      for (auto dbits = 1u; dbits < 16u; ++dbits) {
```

Comment:
Confirming, `[1, 16)`? Not `[0, 16)`?

Reply:
Correct. dbits=0 would generate a zero divisor forever.